### DISCUSSION TITLE
Sharing view state question

### DISCUSSION THREAD

**Discussion by cornhundred** [visgl/deck.gl#7696] [Q&A] [answered]
Hi, I am not understanding how I can get view state sharing working across multiple views (I am working on ObservableHQ and not using React and the notebook can be found [here](https://observablehq.com/d/f7d59770121c4b0d)). 

I am experimentint with a heatmap visualization that has row, column, and matrix sections (each is its own view). The row and column views are restricted to 'Y' and 'X' zooming respectively and the matrix view is set to 'all'. I would like to have the behavior where zooming into the matrix view causes 'Y' and 'X' zooming in the row and column views respectively. However, zooming into the matrix area causes the row and column sections to zoom in 'all' dimensions instead of 'Y' and 'X' respectively. 

![heatmap-gif](https://user-images.githubusercontent.com/8352840/221615459-01c0f36f-8fa9-4e6b-8770-dffe26790e00.gif)

If I try to set the zoom or target of the rows from within the matrix section of the onViewStateChange function this does not be have as expected and seems to effect the zoom state of all views.

```
views = [
      new deck.OrthographicView({
        id: 'matrix',
        x: label_width + 'px',
        y: label_width + 'px',
        width: matrix_width + 'px', 
        height: matrix_height + 'px', 
        controller: {scrollZoom: true, inertia: false, zoomAxis: 'all'},
      }), 

      new deck.OrthographicView({
        id: 'rows',
        x: '0px',
        y: label_width + 'px',
        width: '100px',
        height: matrix_height + 'px',
        controller: {scrollZoom: true, inertia: false, zoomAxis: 'Y'},
      }),       

      new deck.OrthographicView({
        id: 'cols',
        x: label_width + 'px',
        y: '0px',
        width: matrix_width + 'px',
        height: label_width + 'px',
        controller: {scrollZoom: true, inertia: false, zoomAxis: 'X'},
      }),         
      
    ]
```

I initialized a view state with values that match the `id`s from each view:

```
ini_view_state = ({
  'matrix': ({
    target: [500, 500, 0],
    zoom: [-1, -1]
  }),
  'rows': ({
    target: [500, 500, 0],
    zoom: [-1, -1]
  }),
  'cols': ({
    target: [500, 500, 0],
    zoom: [-1, -1]
  })
})
```
and the onViewStateChange function looks like:

```
    onViewStateChange: ({viewState, viewId}) => {

      console.log(viewState)

      if (viewId === 'matrix') {

        // console.log('matrix')
        viewState.matrix = viewState;
        viewState.rows = viewState;
        viewState.cols = viewState;

        // // trying to set the target of the rows
        // viewState.rows.zoom = [0, 0]

      } else if (viewId === 'rows') {

        // console.log('rows')
        viewState.matrix = viewState;
        viewState.rows = viewState;
        viewState.cols = viewState;        
        
      } else if (viewId === 'cols'){
        
        // console.log('cols')
        viewState.matrix = viewState;
        viewState.rows = viewState;
        viewState.cols = viewState;
      }
      
      // Update deck
      instance.setProps({viewState: {...viewState}});
    }
```

**Comment by cornhundred** [comment, marked as answer]
This discussion asks a similar question but I'm unsure how to get a minimal example working on ObservableHQ or other service - https://github.com/visgl/deck.gl/issues/5203#issuecomment-736198821

**Reply by Pessimistress** [reply]
`onViewStateChange` is called with the `viewState` of a single view. You need to keep a variable that is the combined view state and call `setProps` with that.
